param/parameterized.py

```diff
@@ -913,6 +926,44 @@ def self_or_cls(self_):
         return self_.cls if self_.self is None else self_.self
 
 
+    def __getitem__(self_, key):
+        inst = self_.self
+        parameters = self_.params() if inst is None else inst.param.params()
+        p = parameters[key]
+        if inst is not None and p.per_instance and key not in inst._instance__params:
+            p = copy.copy(p)
+            p._owner = inst
+            inst._instance__params[key] = p
+        return p
+
+
+    def __dir__(self_):
+        completions = super(Parameters, self_).__dir__()
+        completions += list(self_.params())
+        return completions
+
+
+    def __getattr__(self_, attr):
+        cls = self_.__dict__.get('cls')
+        if cls is None: # Class not initialized
+            raise AttributeError
+
+        try:
+            params = list(getattr(cls, '_%s__params' % cls.__name__))
+        except AttributeError:
+            params = [n for class_ in classlist(cls) for n, v in class_.__dict__.items()
+                      if isinstance(v, Parameter)]
+
```

Comment:
Looks to me like it should be the other way around - check the methods first, then fall back to the parameters. I wouldn't want a method to be shadowed by a parameter of the same name. What would be best would be to check if a parameter has a matching name to an existing method and warn when returning the method, explaining that the requested parameter has been shadowed (and maybe suggest to use `__getitem__` directly instead.

Reply:
``__getattr__`` only gets called if the attribute wasn't otherwise found.

Reply:
Ok. In that case you can't get a warning if there is a name clash.